File: services/alpha-engine/alpha_engine_svc/order_book.py
"""In-memory order book maintained from depth updates.

This is the Python implementation. Will be swapped for a C++ pybind11
module in Phase 6 without changing the interface.
"""

from __future__ import annotations

import logging
from collections import defaultdict
from typing import Any

from quant_core.models import DepthUpdate

logger = logging.getLogger(__name__)


class OrderBook:
    """L2 order book maintained from incremental depth updates.

    Interface contract (preserved when migrating to C++):
        - apply_delta(update: DepthUpdate) -> None
        - best_bid() -> tuple[float, float] | None
        - best_ask() -> tuple[float, float] | None
        - mid_price() -> float | None
        - spread() -> float | None
        - imbalance(levels: int = 5) -> float
    """

    def __init__(self, symbol: str):
        self.symbol = symbol
        self._bids: dict[float, float] = {}  # price -> quantity
        self._asks: dict[float, float] = {}

    def apply_delta(self, update: DepthUpdate) -> None:
        """Apply an incremental depth update to the book."""
        for price, qty in update.bids:
            if qty == 0.0:
                self._bids.pop(price, None)
            else:
                self._bids[price] = qty

        for price, qty in update.asks:
            if qty == 0.0:
                self._asks.pop(price, None)
            else:
                self._asks[price] = qty

    def best_bid(self) -> tuple[float, float] | None:
        """Highest bid (price, quantity)."""
        if not self._bids:
            return None
        price = max(self._bids)
        return (price, self._bids[price])

    def best_ask(self) -> tuple[float, float] | None:
        """Lowest ask (price, quantity)."""
        if not self._asks:
            return None
        price = min(self._asks)
        return (price, self._asks[price])

    def mid_price(self) -> float | None:
        """Midpoint between best bid and best ask."""
        bid = self.best_bid()
        ask = self.best_ask()
        if bid is None or ask is None:
            return None
        return (bid[0] + ask[0]) / 2.0

    def spread(self) -> float | None:
        """Spread between best ask and best bid."""
        bid = self.best_bid()
        ask = self.best_ask()
        if bid is None or ask is None:
            return None
        return ask[0] - bid[0]

    def imbalance(self, levels: int = 5) -> float:
        """Order book imbalance ratio across top N levels.

        Returns a value between -1.0 (all ask pressure) and 1.0 (all bid pressure).
        0.0 means balanced.
        """
        bid_prices = sorted(self._bids.keys(), reverse=True)[:levels]
        ask_prices = sorted(self._asks.keys())[:levels]

        bid_volume = sum(self._bids[p] for p in bid_prices)
        ask_volume = sum(self._asks[p] for p in ask_prices)

        total = bid_volume + ask_volume
        if total == 0.0:
            return 0.0

        return (bid_volume - ask_volume) / total

    def top_bids(self, levels: int = 10) -> list[tuple[float, float]]:
        """Top N bid levels sorted by price descending."""
        prices = sorted(self._bids.keys(), reverse=True)[:levels]
        return [(p, self._bids[p]) for p in prices]

    def top_asks(self, levels: int = 10) -> list[tuple[float, float]]:
        """Top N ask levels sorted by price ascending."""
        prices = sorted(self._asks.keys())[:levels]
        return [(p, self._asks[p]) for p in prices]

File: cpp/src/order_book.cpp
#include "quant_cpp/order_book.h"

#include <algorithm>
#include <cmath>

namespace quant {

OrderBook::OrderBook(std::string symbol) : symbol_(std::move(symbol)) {}

void OrderBook::apply_delta(
    const std::vector<std::pair<double, double>>& bids,
    const std::vector<std::pair<double, double>>& asks
) {
    for (const auto& [price, qty] : bids) {
        if (qty == 0.0) {
            bids_.erase(price);
        } else {
            bids_[price] = qty;
        }
    }
    for (const auto& [price, qty] : asks) {
        if (qty == 0.0) {
            asks_.erase(price);
        } else {
            asks_[price] = qty;
        }
    }
}

std::optional<std::pair<double, double>> OrderBook::best_bid() const {
    if (bids_.empty()) return std::nullopt;
    const auto& [price, qty] = *bids_.begin();  // greatest key (std::greater)
    return std::make_pair(price, qty);
}

std::optional<std::pair<double, double>> OrderBook::best_ask() const {
    if (asks_.empty()) return std::nullopt;
    const auto& [price, qty] = *asks_.begin();  // smallest key
    return std::make_pair(price, qty);
}

std::optional<double> OrderBook::mid_price() const {
    auto bid = best_bid();
    auto ask = best_ask();
    if (!bid || !ask) return std::nullopt;
    return (bid->first + ask->first) / 2.0;
}

std::optional<double> OrderBook::spread() const {
    auto bid = best_bid();
    auto ask = best_ask();
    if (!bid || !ask) return std::nullopt;
    return ask->first - bid->first;
}

double OrderBook::imbalance(int levels) const {
    double bid_volume = 0.0;
    double ask_volume = 0.0;

    int count = 0;
    for (const auto& [price, qty] : bids_) {
        bid_volume += qty;
        if (++count >= levels) break;
    }

    count = 0;
    for (const auto& [price, qty] : asks_) {
        ask_volume += qty;
        if (++count >= levels) break;
    }

    double total = bid_volume + ask_volume;
    if (total == 0.0) return 0.0;
    return (bid_volume - ask_volume) / total;
}

std::vector<std::pair<double, double>> OrderBook::top_bids(int levels) const {
    std::vector<std::pair<double, double>> result;
    result.reserve(std::min(static_cast<size_t>(levels), bids_.size()));
    int count = 0;
    for (const auto& [price, qty] : bids_) {
        result.emplace_back(price, qty);
        if (++count >= levels) break;
    }
    return result;
}

std::vector<std::pair<double, double>> OrderBook::top_asks(int levels) const {
    std::vector<std::pair<double, double>> result;
    result.reserve(std::min(static_cast<size_t>(levels), asks_.size()));
    int count = 0;
    for (const auto& [price, qty] : asks_) {
        result.emplace_back(price, qty);
        if (++count >= levels) break;
    }
    return result;
}

}  // namespace quant
